src/features/tokenGatedRooms/sagas/join.saga.ts

```diff
@@ -0,0 +1,69 @@
+import { TokenGatedRoomAction } from '$/features/tokenGatedRooms'
+import { takeEvery } from 'redux-saga/effects'
+import handleError from '$/utils/handleError'
+import { JoinPolicyRegistryAddress } from '$/features/tokenGatedRooms/utils/const'
+
+import { Contract, providers } from 'ethers'
+
+import * as JoinPolicyRegistry from '$/contracts/JoinPolicyRegistry.sol/JoinPolicyRegistry.json'
+
+import * as ERC20JoinPolicy from '$/contracts/JoinPolicies/ERC20JoinPolicy.sol/ERC20JoinPolicy.json'
+import * as ERC721JoinPolicy from '$/contracts/JoinPolicies/ERC721JoinPolicy.sol/ERC721JoinPolicy.json'
+import * as ERC777JoinPolicy from '$/contracts/JoinPolicies/ERC777JoinPolicy.sol/ERC777JoinPolicy.json'
+import * as ERC1155JoinPolicy from '$/contracts/JoinPolicies/ERC1155JoinPolicy.sol/ERC1155JoinPolicy.json'
+
+import { TokenTypes } from '$/features/tokenGatedRooms/types'
+
+function* onJoin({
+    payload: { roomId, tokenAddress, provider, stakingEnabled, tokenType },
+}: ReturnType<typeof TokenGatedRoomAction.join>) {
+    try {
+        // fetch policy address from registry
+        const signer = new providers.Web3Provider(provider).getSigner()
+
+        const policyRegistry = new Contract(
+            JoinPolicyRegistryAddress,
+            JoinPolicyRegistry.abi,
+            signer
+        )
+
+        const policyAddress: string = yield policyRegistry.getPolicy(
+            tokenAddress,
+            0, // tokenId = 0 for erc20
+            roomId,
+            stakingEnabled
+        )
+
+        // assemble policy contract
+        let policyAbi: any
+        switch (tokenType.standard) {
+            case TokenTypes.ERC20.standard:
```

Comment:
`TokenTypes.ERC*.standard` can be replaced with `TokenStandard.ERC*` now. Here, and in other places.